content/blog/2019-10-28-self/index.md

```diff
@@ -0,0 +1,176 @@
+becomes impossible in presence of dynamic method calls if the type of the
+receiver object is unknown. This is an even bigger problem for Self since
+most of program is a sequence of dynamic calls.
+
+The Self compiler performs two optimizations, message _inlining_ and
+_splitting_ to enable efficient execution. Message _inlining_ works in a similar
+fashion to customized compilation -- If the type of a receiver is know, either
+at initial compilation through a dataflow analysis or at runtime, inline the
+method body at the call location. When type information for an object is lost
+due to control flow splits, _message splitting_ generates specialized code for
+possible receiver types and guards them using type tests. If the type of
+receiver matches, fast code can be executed.
+
+### Programming Environment Support
+
+The Self compiler support _incremental recompilation_ and _source-level
+debugging_ by keeping track of the provenances of various method
+specializations in a map.
+
+Incremental Recompilation occurs when the compiler observes a change in the
+programming environment and invalidates compiled methods associated with the
+affected data. At compile time, the compiler creates a dependency list encoding
+the information required for specialization. For example, if a method was
+inlined from a clone's parent, updating the parent pointer or the method body
+in the parent should invalidate the specialized code. Since the compiler
+selectively invalidates code, methods that weren't modified can still execute
+with specialized code.
+
+Source-level Debugging requires language support and can often cause compiler
+writers to forgo obvious optimizations. For example, the Chrome V8 team decided
+to [forgo][v8-no-tail] tail call optimization due to due concerns about being
+unable to reconstruct stack frames. The programming environment needs to be
+able to step through specialized code as if it was going through the normal
+method call chain. Self appends debugging information to each compiled method,
+allowing the environment to reconstruct the state of the stack.
+
+## A SELFish Evaluation
+
+The Self compiler was written in 33,000 lines of C++ code and ran on the Sun-3.
+The authors wrote 9,000 lines of Self code to implement the object hierarchy
+and prototype graphical user interface.
+
+The quantitative evaluation compares Self running times to that of the fastest
+Smalltalk-80 compiler at the time by translating the Stanford Integer
+Benchmarks. The authors both did a straightforward transliteration of Self programs
+(marked `Self`) and also rewrote the benchmarks in idiomatic Self (marked `Self'`).
+
+<img src="eval.png" alt="drawing" width="400"/>
+
+On average, Self was 2-3x faster than the Smalltalk implementation and 4x slower
+than the original C programs.
+
+The authors also describes a new metric to measure the performance of an object
+oriented programming language - _Millions of Messages per Second (MiMS)_ which
```

Comment:
Elsewhere, you have used ` -- ` to separate thoughts, rather than just a single hyphen. Consider sticking to that or just using commas here.